src/gui/qgsmapcanvas.cpp

```diff
@@ -778,8 +796,26 @@ QgsRectangle QgsMapCanvas::imageRect( const QImage& img, const QgsMapSettings& m
 
 void QgsMapCanvas::mapUpdateTimeout()
 {
-  const QImage& img = mJob->renderedImage();
-  mMap->setContent( img, imageRect( img, mJob->mapSettings() ) );
+  QImage img = mJob->renderedImage();
+  magnifyImage( img );
+
+  mMap->setContent( img, imageRect( img, mSettings ) );
+}
+
+void QgsMapCanvas::magnifyImage( QImage &img )
+{
+  if ( mMagnificationLevel != 1.0 )
+  {
+    img = img.scaled( img.width() * mMagnificationLevel,
```

Comment:
I think we still need the viewport extent changes. Otherwise zooming in will make labels near the edge of the canvas suddenly disappear when they should be visible.